// @okta/okta-angular/esm2015/src/okta/components/callback.component.js
/*
 * Copyright (c) 2017-Present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { Router } from '@angular/router';
import { OktaAuthService } from '../services/okta.service';
let OktaCallbackComponent = class OktaCallbackComponent {
    constructor(okta, router) {
        this.okta = okta;
        this.router = router;
    }
    ngOnInit() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                // Parse code or tokens from the URL, store tokens in the TokenManager, and redirect back to the originalUri
                yield this.okta.handleLoginRedirect();
            }
            catch (e) {
                this.error = e.toString();
            }
        });
    }
};
OktaCallbackComponent = tslib_1.__decorate([
    Component({
        template: `<div>{{error}}</div>`
    }),
    tslib_1.__metadata("design:paramtypes", [OktaAuthService, Router])
], OktaCallbackComponent);
export { OktaCallbackComponent };
